src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java

```diff
@@ -0,0 +1,251 @@
+
+import static frc.robot.Ports.SwerveDrive.*;
+
+/**
+ * The Swerve Subsystem calculates all the mathematical operations and controls for the Swerve Modules.
+ */
+public class SwerveDrive extends SubsystemBase {
+
+    private static double[][] dynamics = new double[8][3];
+    private SwerveModule[] swerveModules = new SwerveModule[4];
+
+    // calculates the distance from the center of the robot to the wheels
+    private static double Rx = Constants.SwerveDrive.ROBOT_WIDTH / 2;
+    private static double Ry = Constants.SwerveDrive.ROBOT_LENGTH / 2;
+
+    // the sign vectors of Rx and Ry
+    private static double[] signX = {1, 1, -1, -1};
+    private static double[] signY = {-1, 1, -1, 1};
+
+    private static boolean isFieldOriented;
+
+    public SwerveDrive(boolean isFieldOriented, boolean testMode) {
+        createInverseMatrix();
+
+        if (!testMode) {
+            swerveModules[0] = new SwerveModule(0, DRIVE_MOTOR_FRONT_RIGHT, ANGLE_MOTOR_FRONT_RIGHT, FRONT_RIGHT_INVERTED,
+                    Constants.SwerveModule.ANGLE_PIDF, Constants.SwerveModule.DRIVE_PIDF);
+
+            swerveModules[1] = new SwerveModule(1, DRIVE_MOTOR_FRONT_LEFT, ANGLE_MOTOR_FRONT_LEFT, FRONT_LEFT_INVERTED,
+                    Constants.SwerveModule.ANGLE_PIDF, Constants.SwerveModule.SLOW_DRIVE_PIDF);
+
+            swerveModules[2] = new SwerveModule(2, DRIVE_MOTOR_BACK_RIGHT, ANGLE_MOTOR_BACK_RIGHT, BACK_RIGHT_INVERTED,
+                    Constants.SwerveModule.SICK_ANGLE_PIDF, Constants.SwerveModule.DRIVE_PIDF);
+
+            swerveModules[3] = new SwerveModule(3, DRIVE_MOTOR_BACK_LEFT, ANGLE_MOTOR_BACK_LEFT, BACK_LEFT_INVERTED,
+                    Constants.SwerveModule.ANGLE_PIDF, Constants.SwerveModule.DRIVE_PIDF);
+        }
+
+        this.isFieldOriented = isFieldOriented;
+    }
+
+    /**
+     * sets the wheels of the robot to the calculated angle and speed
+     *
+     * @param forward  the Y value of the joystick
+     * @param strafe   the X value of the joystick
+     * @param rotation the rotation Z of the joystick
+     */
+    public void holonomicDrive(double forward, double strafe, double rotation) {
+
+        double[] robotHeading = getRobotHeading(strafe, forward, rotation, -Math.toRadians(Robot.gyro.getAngle()));
+
+        double[] velocities = calculateWheelVelocities(robotHeading);
+        double[] polar;
+        double[][] controls = new double[4][2];
+
+        // converts the cartesian velocities to polar and transfers them to a control matrix
+        for (int i = 0; i < 4; i++) {
+            polar = Utils.cartesianToPolar(velocities[2 * i + 1], velocities[2 * i]);
+            controls[i][0] = polar[0];
+            controls[i][1] = polar[1];
+        }
+
+        // feeds the corresponding control to each wheel
+        for (int k = 0; k < 4; k++) {
+            swerveModules[k].setSpeed(controls[k][0]);
+            swerveModules[k].setAngle(controls[k][1]);
+        }
```

Comment:
Why not use this in the above for loop? (then you can also remove the controls 2d array)

Reply:
I did this on purpose to make the code more understandable and "highlight" the controls and if needed it will be easy to check their values. If you think this is unnecessary I can change it.